Phil/cpv2

psFried opened the pull request:
**Description:**

Rolls up whole bunch of work from both @jgraettinger and myself related to:

- publications
    - builds
    - validation
    - activation
- inferred schema handling
- materialization schema evolution (`onIncompatibleSchemaChange`)
- controllers :sparkles: 

I recommend reviewing this commit by commit, and tried to write helpful commit messages. Here's a high-level summary:

Reframe validation to be an operation over a `tables::LiveCatalog` (a snapshot from the control plane) and a `tables::DraftCatalog`, which is able to represent deletions.

Updates the publications handler to:
- work with the new `validation` crate changes
- use optimistic locking instead of holding row-level locks
- setup for a future `publish` function that can be invoked by controllers
- use a more minimal spec expansion
- trigger controller runs after a successful publication

Introduces controllers, which are background automations that are tied to specific live specs. Controllers are responsible for:

- updating shard and journal specs in the data plane, in response to a publication
- keeping specs up-to-date with respect to their dependencies
- updating collection inferred schemas (inferred schemas are now added directly to the collection spec)
- updating materailizations in response to incompatible schema changes
- updated materializations in response to changes of their `sourceCapture`
- maintaining a status that can provide some basic information about their operation and any errors that occur

**Workflow steps:**

The control plane should work more or less as it has in terms of user-facing operations, with a few exceptions:

- You can now delete and re-create collections and tasks with the same name :tada: 
- Inferred schemas now work with derivations
- Inferred schemas now get baked into the collection model. This means that you can technically now overwrite the published inferred schema with flowctl. This might be easy to do, since it's hard to know if the inferred schema has changed, and your local spec is now stale.
- Materializations now obey the `onIncompatibleSchemaChange` property (top-level, or can be overridden on a per-binding basis)
- Users will need to update flowctl in order for things to work with the new control plane. One reason for this is that the `spec_type` column on `live|draft_specs` is now non-nullable.
- Activations now happen asynchronously, after the publication itself has already succeeded. Thus there is not a possibility of a slight delay after publication before the new shard status can be shown on the "details" page. This means a user can publish a task that was failing, and then get redirected to the details page and possibly be confused by the shard status still showing as failed. One way to address this might be to have the details page query shard status a little more frequently. We also have the information available in `controller_jobs.status->activation->last_activated`, which can be compared agains `live_specs.last_pub_id` to see if the activation has been completed yet.
- We no longer return `JobStatus::TestsFailed` when a test run fails. Instead it just returns `BuildFailed` (the can be fixed but I wasn't sure if it was important).

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

I'd appreciate any time you have to test this yourself locally, and report any issues. I'd like to continue testing, myself, particularly on the transition from the old version, and I expect I'll need to push some updates before this is ready to deploy. I'd like to keep this PR open until we feel this is in a deployable state.

Please reach out by any means with any questions.

### Deployment plan

The basic steps for deploying this are the same as any other control plane PR involving a database migration, except that there's an additional migration to be run _after_ the deployment finishes. This second migration will actually need to be run multiple times, because it will incrementally enable controllers for pre-existing live specs, in batches of 1000 at a time. This gives us a chance to bail out if things aren't working right, instead of just unleashing controllers on all of our precious live specs. The end-to-end high level plan is thus:

- Run the `50_controllers.sql` migration, previous agent versions will continue to run against the new db schema
- Deploy new agents, just using the normal ops repo
- Verify that we can publish a new task and edit an existing one through the UI, verify controller runs in `controller_jobs`
- Run the `51_backfill_controllers.sql` migration, and spot check the controller runs to make sure things look good
- Repeat the previous step until we've backfilled all of them
- Profit


<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/1489)
<!-- Reviewable:end -->

jgraettinger committed:
pending agent changes to be smushed by Phil

jgraettinger committed:
crates/models: add `Id`, `ModelDef`, `expect_pub_id` & `delete`

`Id` is the (existing) foundational identifier of the Estuary control plane.
It's a 64-bit integer that embeds a timestamp and sequence bits,
and is serialized similarly to a MAC address. It corresponds to the
MACADDR8 Postgres type.

`ModelDef` is a common trait of top-level specification models,
and currently enumerates their sources and targets.

`expect_pub_id` allows a model to represent its expected
specification `last_pub_id` within the control plane.

`delete` allows a model to represent its intent to be
deleted upon its next publication.

To allow tests to represent these new fields, introduce models::TestDef
which is a structured mapping that contains the former list of test
steps. A custom serde::Deserialize decodes either the legacy or new
form, while models::TestDef is always serialized in its new form.

Also remove `storageMappings` from models::Catalog.
These will be provided through tables::LiveCatalog.

jgraettinger committed:
crates/tables: DraftCatalog, LiveCatalog, and joins

Broadly update tables to enable representing a potential state
transition and transactional update of the global Estary catalog.

LiveCatalog reflects a partial snapshot of the Estuary catalog,
including current specification models and built specs,
and tagged with the `last_pub_id` of their respective _last_
publications.

DraftCatalog reflects a desired change to be applied to the catalog
in terms of specification models, optionally with `expect_pub_id` to
which those models should be applied.

Validations (to be renamed BuiltCatalog) reflects a validation outcome,
and contains metadata as to whether the specification is updated,
its model, its previous built specification, and the `expect_pub_id`
which was currently effective as-of this build.

BuiltCatalog thus encodes an optimistic transaction which can be applied
by validating all `expect_pub_id` priors and then updating models and built
specifications within the global control plane.

Also add new routines to facilitate joins across tables, as we have more
of them and now need to join across them.

jgraettinger committed:
crates/sources: update for `tables` and `models` changes

These changes are largely mechanical.

One interesting implementation choice is that `expect_pub_id` and
`delete` are handled at load time by immediately discarding the internal
field of the model and instead setting it on its tables::Draft* row.

This means serialized models within a tables::DraftCatalog or the
control plane will _never_ have these model fields set.

Then, when mapping a tables::DraftCatalog into various models::Catalog
for serialization, this mapping is un-done for `expect_pub_id`.

No attempt is made to round-trip a tables::Draft* deletion row into a
non-deleted specification model with `delete: true` (there also isn't
a use case that I'm aware of).

jgraettinger committed:
crates/assemble: update for `tables` and `model` changes

These are mostly mechanical updates.

One interesting change is the introduction of `journal_name_prefix` and
`shard_id_prefix` in routnes for building partition templates, recovery
log templates, and shard templates.

Previously we were simply using the collection or task name. We'll use
the indirected arguments to support perpetuating a prefix selected by a
previous catalog build.

`shard_id_prefix` also introduces the pattern of suffixing a candidate
Shard ID prefix with the `pub_id` of the publication which initially
creates it.

jgraettinger committed:
crates/validation: reframe in terms of state transition

Previously, validation was an operation that took (only) a draft catalog
of upserted models and produced built tables::Validations.
It used a ControlPlane trait to resolve limited information from the
control plane (such as referenced collections and inferred schemas).

Instead, reframe validation to be an operation over a
tables::LiveCatalog (a snapshot from the control plane) and a
tables::DraftCatalog, which is able to represent deletions and
CAS `last_pub_id` values.

The job of validation is to verify the _transition_ from current live
specifications to the drafted model, and produce a tables::Validations
(to be renamed tables::BuiltCatalog) outcome, which encodes an
optimistic transaction to apply to the control plane (assuming there are
no errors).

Two immediate improvements are that we're now able to handle deletions
and re-creations of existing specifications, because we generate and
later carry-forward unique journal and shard template prefixes that
reflect the `pub_id` of their original creation.

We also now thread through previous specs into connector Validate RPCs,
which has been a todo for a while now.

The ControlPlane trait is removed.

We do intend to push additional validations down into this crate from
`agent`, but... not yet.

jgraettinger committed:
crates/build: update for validation/tables/models changes

These are pretty mechanical.

We do formalize `pub_id` and `build_id` as separate identifiers which
must be known and passed into a validation / catalog build.

`pub_id` is the publication ID within the control plane -- a revision
that is used for transactional checks and eventualy convergence of
global catalog specifications.

`build_id` uniquely identifies a write-once build database which stores
built specifications for the runtime.

jgraettinger committed:
crates/flowctl: update for `models` and `tables` changes

A lot of mechanical changes due to `tables` and `models` updates.

Generally, routines that previously worked with `models::Catalog` are
updated to work with `tables::DraftCatalog`

Implement the new tables::CatalogResolver trait for fetching a
tables::LiveCatalog.

Remove the ability to list deleted specs, as we do not intend to
retain tombstones for these going forward.

Fix a long-standing bug in tailing publication logs, which would
previously re-emit the _entire_ set of logs on every poll.

Update liberally to models::Id, and make SpecType required in fetched
models. We intend to use _only_ `spec` NULL's for representing drafted
deletions going forward.

Stop disabling UUID placeholders when running `flowctl preview`, as
we've had repeated issues where this difference in behavior between
`flowctl` and the runtime tripped up schema validation when
transitioning a connector from development to production.

jgraettinger committed:
crates/flowctl: update for `models` and `tables` changes

A lot of mechanical changes due to `tables` and `models` updates.

Generally, routines that previously worked with `models::Catalog` are
updated to work with `tables::DraftCatalog`

Implement the new tables::CatalogResolver trait for fetching a
tables::LiveCatalog.

Remove the ability to list deleted specs, as we do not intend to
retain tombstones for these going forward.

Fix a long-standing bug in tailing publication logs, which would
previously re-emit the _entire_ set of logs on every poll.

Update liberally to models::Id, and make SpecType required in fetched
models. We intend to use _only_ `spec` NULL's for representing drafted
deletions going forward.

Stop disabling UUID placeholders when running `flowctl preview`, as
we've had repeated issues where this difference in behavior between
`flowctl` and the runtime tripped up schema validation when
transitioning a connector from development to production.

jgraettinger committed:
crates/labels: new routines for partitions, and refactor

Add and update routines for encoding and decoding labels of collection
partitions.

Routines and tests generally match those of the Go `labels` package.

Refactor the `labels` crate into separate `shard` and `partition`
sub-modules.

jgraettinger committed:
crates/gazette: add a basic shard::Client with list & apply support

Also add apply RPC to journal::Client.

jgraettinger committed:
crates/build: minor refactors to support tests

Make build::Output fields public, and add NoOpCatalogResolver
to support usages of build::managed_build() in test cases.

This gets us parity with Go test infastructure we've been using in test
cases to perform integrated tests with Flow catalogs, without an
integrated control plane.

jgraettinger committed:
crates/activate: converge data-plane specs to desired states

`activate` implements details of listing and inserting / updating /
deleting data-plane specs to bring them into alignment with a desired
control-plane state.

It's replacing older Go infrastructure for performing task activations.

psFried committed:
activate: Unassign failed shards during activation

Brings the `activate` crate in line with the behavior of `flowctl-go api activate`,
which unassigns failed shards after updating shard specs. This ensures
that failed shards will get restarted after a new spec has been
published.

psFried committed:
models: add `onIncompatibleSchemaChange` and beef up `ModelDef` trait

Adds `onIncompatibleSchemaChange` to materialization specs. This
property allows materializations to define how to respond to incompabile
schema changes. The implementation of that behavior will come in a
subsequent commit.

Adds some more functions to the `ModelDef` train in support of
generrically determining the dependencies of specs. This is also
just setting the stage for work in a future commit, where these
functions will be used as part of publications and controllers.

psFried committed:
models: add `IdGenerator`

Adds the `IdGenerator` to the models crate, which generates new unique
`Id`s. This sets up future work for controllers, which will need to
generate their own publication ids (rather than generating the ids in
the database). The implementation here closely follows the one in
`02_flowid.sql`, with a couple of differences:

- `IdGenerator` allows using any shard id (the sql version always uses
  `0`)
- `IdGenerator` resets the sequence counter to 0 whenever the timestamp
  increases, in order to ensure that `Id`s always increase
  monotonically.

psFried committed:
local: publish ops collections separately

This sets up for a forthcoming change to how publications work, which
will require that ops collections exist as `live_specs` in the control
plane, rather than being generated as part of the build. This breaks up
the ops catalog into separate parts, with `base-collections.flow.yaml`
only containing the collections themselves. The reason for this is that
the ops materializations require the appropriate `connector_tags` row
to have been processed, and publishing the entire ops catalog would thus
race against the `connector_tags` job. There's probably better ways to
do this, but I deemed this low-effort solution "goodenuf", since there
are loads more changes coming soon with federated data planes, which
will likely also impact how we manage ops catalogs locally.

psFried committed:
tables: add functions for working with catalogs

Adds various helper functions for working with draft, live, and built
catalogs. This sets up more work in a subsequent commit for changing
publications over to use the new `tables` types, and for implementing
controllers.

psFried committed:
agent: publications redux

Updates the publications handler to:
- work with the new `validation` crate changes
- use optimistic locking instead of holding row-level locks
- setup for a future `publish` function that can be invoked by
  controllers
- use a more minimal spec expansion
- trigger controller runs after a successful publication

Also introduces the controllers database migration, which is needed in
order for the new publications handler to work properly.

**`validation` crate changes**

The `validation` crate has been updated to operate in terms of
`DraftCatalog` and `LiveCatalog`, which are both expected to be fully
populated prior to validating (i.e. `validate` no longer dynamically
fetches live specs). Other relevant changes are that the output of
`validate` now indicates whether each spec was changed, and if it's an
insert, update, or deletion. Publications have been updated to work with
these changes, which results in a much cleaner factoring of the
publication logic.

**Optimistic locking**

The use of optimistic locking is intended to help with scaling.
Previously, agents would lock `live_specs` rows at the start of a
publication, and continue to hold the locks during the lengthy process
of validation, until the transaction could either be committed or rolled
back. Now, it no longer locks any rows until the build has already been
deemed successful and the published changes are ready to commit.

The overall publication process has been broken up into three basic parts:
- `Publisher::build(...) -> Result<UncommittedBuild>`, which validates a
  draft and returns an `UncommittedBuild`, which can then be checked for
  errors and committed if there are none. This function acquires no
  database locks, and does not modify the database at all.
- `commit(uncommitted_build) -> Result<PublicationResult>`, which
  commits a successful `UncommittedBuild`, persisting all the results to
  the database. This will return an `ExpectPubIdMismatch` status if any
  drafted specs have been modified before the commit.
- The handler's `process` function, which calls `build` and `commit`,
  and retries if it encounters an `ExpectPubIdMismatch`.

**Future `publish` function**

A subsequent commit will introduce a `publish` function that's callable
by controllers to immediately publish a given draft. This new factoring
is intended to make it easy to write such a function. The handler's
`process` loop serves as an example, and is actually quite simple.
Another future aspiration is to expose a `/publish` HTTP endpoint, which
would also be enabled by this change.

**Spec expansion**

Our previous spec expansion logic was known to be overly broad, which
resulted in unnecessary lock contention, and sometimes confusing errors
when validation fails for expanded tasks that the user didn't expect to
be included in the build. The new spec expansion logic is much more
minimal, and only expands to tasks that read from or write to a
collection that's been drafted by the user, and which the user has
`admin` capability to.

This more minimal expansion logic is enabled by the changes to
`validation`, which now uses a separate `LiveCatalog` to access
information about referenced specs that may or may not have been
drafted. Thus spec expansion is now technically optional from the
perspective of `validation`, as long as we pass the required specs as
part of the `LiveCatalog`. This lets us validate the subset of specs
that make the most sense for the given scenario. For the publications
handler, the spec expansion logic is intended to prevent users from
publishing changes that could be incompatible with connected tasks and
tests, but the bounds of that no longer cross permission boundaries.
Take for example, an `orgA/collection` that's materialized by
`orgB/materialization`. If `orgA/` doesn't have `admin` capability to
`orgB/`, then the materialization validation will no longer prevent an
`orgA/` user from publishing an incompatible schema change to the
collection.

**Trigger controller runs after a successful publication**

The `controllers.sql` migration begins to introduce the concept of
controllers, which are background automations that relate to a specific
live spec. Controllers are not directly involved in publications, per
se, but they do need to be notified whenever their spec is published.
The most relevant issue at this point is activations, which are no
longer handled as part of the publication. Committing a publication will
no longer modify any shard or journal specs, and instead this is
expected to be handled asynchronously by controllers.

One reason for this is to address a long standing issue related to
failed activation calls. Previously, the publications handler would
update all shards and journals prior to committing the publication
transaction. If a shard or journal update failed, then the publication
transaction would rollback, but there was no rollback of the
shard/journal changes. So we could be left in a state where the control
plane has an outdated view of the specs that are actually being run in
the data plane. The other reason for this change is simply to make
committing the publication faster, and minimize the duration of any held
locks.

The end result is that controllers, in a subsequent commit, will handle
all activations.

psFried committed:
agent: introduce controllers

This is a big commit, which introduces the initial implementation of
controllers. Controllers are background automations that are tied to
specific live specs. As of this commit, controllers are responsible for:

- updating shard and journal specs in the data plane, in response to a
  publication
- keeping specs up-to-date with respect to their dependencies
- updating collection inferred schemas
- updating materailizations in response to incompatible schema changes
- updated materializations in response to changes of their
  `sourceCapture`
- maintaining a status that can provide some basic information about
  their operation and any errors that occur

To do all these things, they rely on the new `ControlPlane` trait, which
provides APIs for fetching specs, publishing, activating, etc.

And to top it all off, there's some new `integration_tests`, which test
publications and controllers, and how they interact.

**Basic operation**

Controllers are invoked by the `ControllerHandler`, which works just
like all the other handlers, by dequeuing jobs based on the new
`live_specs.controller_next_run` column. This is allowed to be `null`,
which means that there's no run scheduled, and they can also be
significantly in the future. Controllers have some state that's stored
in the `controller_jobs` table, and gets updated with the results of
each run.

**Updating shard and journal specs in the data plane**

Controllers are responsible for "activation", which includes all updates
to task shards and journal specs. Activations are the first thing that
controllers do, and a failure to activate will short circuit further
controller operations. Failures will be retried with a backoff.

The last activated publication id is recorded as part of the status.

**Keeping specs up-to-date with respect to their dependencies**

This is an essential part of how controllers work, and how the control
plane works end-to-end. Recall that publications no longer _need_ to
fully expand the draft, so it's valid now to publish a single spec in
isolation. Recall also that publication ids always increase monotonically.
Whith that in mind, I'll describe the general operation of controllers,
which is based on the `last_pub_id` of `live_specs`.

When a spec is published, the publications handler will set the
`controller_next_run` to `now()`, which schedules an immediate run of
that controller. When the controller runs, it sees that its
`last_pub_id` has increased, and so calls `notify_dependents`, which set
the `controller_next_run` of all dependent specs, triggering them to
run. When _those_ controllers run, they check all of their dependencies
and see that the `last_pub_id` of a spec that they depend on is now
greater than their own `last_pub_id`, and thus determine that they need
to publish their own spec. When the publish their own spec, they use
that same publication id. So after their publication succeeds, the
largets `last_pub_id` from among the dependencies will become the new
`last_pub_id` of the dependent spec. In this way, the publication of one
spec will propagate through the entire graph of dependent specs. Because
specs are published at the `last_pub_id` of their dependencies, it means
that dependency cycles will not create an infinite loop of controller
runs.

**Updating collection inferred schemas**

Collection controllers check the latest available inferred schema, and
they publish the collection if it's changed. A major change that goes
along with is that inferred schemas are now directly inlined into the
collection model. The motivation for this was to make inferred schema
changes more visible.

When the controller publishes the inferred schema, the publication does
_not_ include tasks that read from or write to the collection. It's just
that one isolated collection. The controllers of any tasks that depend
on the collection will instead be notified after the successful
completion of the publication, and they can respond as they see fit. The
basic idea is that the inferred schema _has_ changed, and so was should
not delay in having the spec reflect those changes. Consumers are free
to respond however they see fit, but they should never be able to
prevent the publication of the inferred schema.

The md5 of the inferred schema and the time of the last update are
recorded as part of the status.

**updating materailizations in response to incompatible schema changes**

Materialization controllers are notified after any publication of a
collection that they read from. They respond by publishing the
materialization. If this publication fails due to "unsatisfiable"
constraints, then the controller will respond based on the
`onIncompatibleSchemaChange` properties in the materialization spec, and
attempt to publish the updated spec. The default is to always just
increment `backfill` counters. The `incompatible_collections` from the
failed publication can be seen in the publication history on the
controller's status.

**updating materializations in response to changes of their `sourceCapture`**

Similarly, materializations "depend on" their `sourceCapture`s, and so
they are notified after each publication of it. The controller responds
by adding any missing bindings and publishing the new spec.

**maintaining a status**

The `controller_jobs` table contains the state of each controller, which
includes general things like `error` and `updated_at`, along with a
`status` column containing the serialized internal state of the
controller. The `status` serves both as the internal state of the
controller, and as a mechanism to communicate relevant information to
users and operators. It's likely that the shape of the `status` will
evolve over time, so there's a `controller_version` column that gives us
the ability to roll out new controller versions gracefully and upgrade
old statuses to new.

**Integration tests**

Controllers turn the control plane into a dynamic system, which makes it
difficult to test things only in isolation. The `TestHarness` was
introduced to enable writing tests of high-level scenarios like schema
evolution, which tend to play out over the course of multiple
publications and controller runs. The tests introduced so far are
intended as a bare minimum of coverage, which should grow, probably
pretty quickly at first.

psFried committed:
control-plane: add spec_type to sql test setup

The `spec_type` column was made `not null`, so this just makes the tests
work with that.

psFried committed:
sources and activate: update test snapshots

psFried committed:
assemble: format build id without colon separators

The colon separators appear in ids by default now, which caused the
assemble crate to output invalid shard labels because colons are not
allowed in label values. Updates assemble to format ids without the
colons.

psFried committed:
update test snapshots again

jgraettinger committed:
labels: add_value must preserve the sorted order of values

Determine the insertion index for the added value, and suppress adding
a duplicate value for a given label name.

psFried committed:
activate: don't unassign shards after deletion

Skips shard unassignment when deleting shards. It doesn't make sense to
try to unassign failed shards right after deleting them. This would only
sometimes result in an error because it races against the teardown of
the shard.

psFried committed:
agent: always stringify build ids without colons

Colons are not allowed in shard and journal names or labels, and they
were sneaking in there because our default `Display` impl for `Id`
includes them as visual separtors between each byte. This updates all
the call sites that convert `Id`s to strings to instead use the
alternate format, which excludes the colons.

jgraettinger commented:
>  This means that you can technically now overwrite the published inferred schema with flowctl. This might be easy to do, since it's hard to know if the inferred schema has changed, and your local spec is now stale.

`flowctl` now adds `expectPubId` to any live specs it pulls down from the control plane. `expectPubId` of zero means "must not exist".

This is to prevent a user from inadvertently overwritting an updated live spec by publishing a stale spec from their checkout. They'll get an error because expectPubId will disagree.

They'd have to explicitly remove the expectPubId field to denote that they _want_ their local version to take precedence.

`flowctl` tries not to publish a spec which isn't actually changing, so they should be good to leave unmodified specs as-in in a local checkout while publishing a managed spec over-and-over.

psFried committed:
update build ids in go tests

Build ids are now required to be a valid `models::Id`. Also, publication
ids now get incorporated into shard and journal names. This updates all
the go tests to use valid build ids, and to expect the publication ids
when verifying shards and journals.

Also removes `storageMappings` from flow catalog yamls, since they are
no longer supported there.

psFried committed:
update build_id for end-to-end tests

psFried committed:
agent: delete inferred schemas with live_specs

Delete inferred schemas when hard-deleting `live_specs` rows.
This prevents the inferred schema from a previous version of a
collection from being used when re-creating a new collection with the
same name.

jgraettinger commented on crates/tables/src/built.rs:
```diff
@@ -232,3 +234,88 @@ impl BuiltRow for crate::BuiltTest {
+            .chain(self.built_tests.iter().map(|r| r.catalog_name().as_str()))
+    }
+
+    pub fn spec_count(&self) -> usize {
```
`all_spec_names().count()` ?

psFried replied:
done

jgraettinger commented on crates/tables/src/draft.rs:
```diff
@@ -1,6 +1,73 @@
+            .chain(self.tests.iter().map(|t| t.catalog_name().as_str()))
+    }
+
+    pub fn to_models_catalog(&self) -> models::Catalog {
```
this isn't called that I can find

psFried replied:
removed

jgraettinger reviewed (commented):
Some initial feedback. I've gotten up to (but haven't started) agent: publications redux

psFried commented on crates/models/src/lib.rs:
```diff
@@ -42,18 +44,245 @@ pub use references::{
+        }
+    }
+
+    fn sources(&self) -> impl Iterator<Item = &Source> {
```
ah, that needs a comment, and probably should be replaced with a `panic!`. The reason for this, and for `reads_from/writes_to`, is that this union type doesn't work well with `impl Iterator<Item = &_>`. `AnySpec` can't easily abstract over the different concrete types of the iterators returned by the various `models::_Def`. It can be done, but didn't seem worth it just yet, because nothing actually calls those functions on this type.

Ideally, we'd unify `sources` and `reads_from`, and so I was also thinking perhaps the borrowed iterator version might go away. I think even that unification can wait a bit until we get a better idea of what else we might do with controllers, though. If, for example, we start adding more types of dependencies, then it might be better still to go with something like `fn dependencies(&self) -> Vec<(String, FlowType)>`.

So I'll propose adding an explicit `unimplemented!` note there for now and kick the can a little.

psFried committed:
agent: make migration compatible with old agents

Updates the controllers migration to allow old agent processes to run
and publish sucessfully during the transition period. This pretty much
just means allowing `spec_type` to remain nullable, and subsequently
dealing with null `spec_type` values in the new agent. Specs having no
spec type will simply be deleted by the new agent version.

Also filters out `test` specs before deleting things in the data plane,
since tests don't exist in the data plane.

jgraettinger commented on crates/agent-sql/src/live_specs.rs:
```diff
@@ -0,0 +1,162 @@
+
+/// Queries for all non-deleted `live_specs` that are connected to the given `collection_names` via
+/// `live_spec_flows`.
+pub async fn fetch_expanded_live_specs(
```
This routine is run only on user-initiated publications, as I understand :+1:

One note is we still need the recursive resolution behavior of resolve_expanded_rows() as well, but now we only need it to pull in live and unchanged derivations which are attached to any drafted tests.

We still need to recursively walk such derivations to pull out the full data flow cycle being put under test.

psFried replied:
I wasn't thinking we needed to do that. The expansion logic here applies to the draft, which will subsequently go though another round of resolution to pull in all the collections referenced by tasks/tests in the draft. So this function expands from published collections to any tests/tasks that depend on them, and then the subsequent `live_specs` resolution will pull in collection specs that those tests need in order to run. I really should have more explicit tests for this, which might help to illustrate that better. But I'm not 100% sure I've got your point. Are you saying that we should try to expand through an arbitrary number of derivations until we get to the captures/materializations "edges"? _That_ would be pretty different from what's here.

psFried replied:
BTW the best spec expansion test we have right now is [`user_publications.rs`](https://github.com/estuary/flow/blob/2570da4f8b7b028da9f45f058159189f2b06377e/crates/agent/src/integration_tests/user_publications.rs#L167).

psFried replied:
We discussed this on VC and I created #1494 to track the issue

jgraettinger commented on crates/agent/src/publications/status.rs:
```diff
@@ -0,0 +1,193 @@
+    KeyChange,
+    /// One or more collection partition fields in the draft differs from that of the live spec.
+    PartitionChange,
+    /// A live spec with the same name has already been created and was subsequently deleted.
```
nit: remove?

psFried replied:
I'd been leaving all the existing publication status fields, just in case we needed to deserialize an old status for some reason. That's now seeming overly cautious, though, so I just removed this and the `linked_materializations` stuff.

jgraettinger commented on supabase/migrations/50_controllers.sql:
```diff
@@ -0,0 +1,127 @@
+begin;
```
(meta: stopping here for now)

jgraettinger reviewed (commented):
More comments! thanks

psFried commented on crates/agent-sql/src/publications.rs:
```diff
@@ -1,10 +1,165 @@
+                else
+                    array(select json_array_elements_text(reads_from))
+                end,
+                case when json_typeof(reads_from) is null then
```
:grimacing: geeze good catch. I'm ... gonna try adding a test that covers this

psFried replied:
Fixed and added some test assertions in [this commit](https://github.com/estuary/flow/pull/1489/commits/b998cd3796035f7077cf5f8a53dafd21312b83cb)

psFried committed:
always stringify ids without colons

Updates the `Display` impl of `models::Id` to never use the colon
separators. Removing the separators allows terminal users to
double-click on the text to select the whole id. And removing the colons
keeps the id width shorter, leaving more space for human-readable info.

psFried committed:
agent: set reads_from/writes_to to null when unused

Updates the agent to set `live_specs` columns `reads_from/writes_to`
to `null` for specs that don't read or write. For example, captures
should always have `reads_from = null`. Collections will have
`reads_from` and `writes_to` both set to null unless it's a derivation,
in which case `reads_from` will be non-null.

Fixing this revealed another bug in the `update_live_specs` sql query,
which was checking the nullability of `reads_from` when setting the
`writes_to` value. I added some assertions to `test_user_publications`
to verify that these are getting set properly.

psFried committed:
agent: use unique tenant names in quotas tests

Updates the quotas integration tests to use unique tenant names, to
avoid colliding with the agent-sql quotas tests.

jgraettinger commented on supabase/migrations/50_controllers.sql:
```diff
@@ -0,0 +1,127 @@
+create table controller_jobs (
+    -- The name of the live spec that this pertains to
+    live_spec_id flowid not null references live_specs (id) on delete cascade,
+    -- The version of the controller that last updated this row. Used to identify controllers to run
```
🤔 how does this impact query performance of the controller? If we're querying for rows where the controller version is less than our current version, that would need to be indexed or it'll require a full table scan (i haven't gotten to this code yet; or is it an output-only column?).

If the goal is to be able to trigger controllers on software releases, it seems like we could as-easily do that through a migration that tickles `controller_next_run` on `live_specs` -- that might even be better because we could set up a schedule that staggers / slow-rolls the controller runs.

psFried replied:
We talked about this on VC and [this commit](https://github.com/estuary/flow/pull/1489/commits/c33c583b4242cfdc3afa6870052aafe9d4bd4dea) introduces a better approach for the migration. Also edited the pr description to add an overall deployment plan.

jgraettinger commented on crates/activate/src/lib.rs:
```diff
@@ -524,24 +536,41 @@ mod test {
-    #[tokio::test]
-    async fn fixture_subtests() {
-        let source = build::arg_source_to_url("./src/test.flow.yaml", false).unwrap();
+    async fn managed_build(source: url::Url) -> build::Output {
```
this was to get rid of build::managed_build ? Was it the last remaining usage?

This test introduced a pattern we've had in Go for a while, of doing an actual catalog build in service of establishing a test case fixture. We do want to keep that path as easy to walk as possible, perhaps by adjusting managed_build's interface if there are no other non-test usages.

psFried replied:
Yeah, this was to get rid of `build::managed_build`, as this was the last remaining usage. We're agreed that we want to keep an easy way to do builds of test fixtures. Perhaps it'll make sense to pull this back over into `build::managed_build` with a `#[cfg(test)]`, but there's also the option of using `Publisher::build` in tests. Both would have a few minor peculiarities, and I feel like I'd have a stronger opinion after seeing the needs of the tests, which is why I stuffed this in here for now.

jgraettinger reviewed (commented):
A few more comments, about to dig into `agent: introduce controllers`.

psFried committed:
models: consolidate ModelDef::reads_from/writes_to

Consolidates the `reads_from` and `writes_to` impls into the `ModelDef`
trait instead of having each type define them separately.

jgraettinger commented on crates/agent-sql/src/controllers.rs:
```diff
@@ -0,0 +1,172 @@
+            order by controller_next_run asc
+        ),
+        needs_upgrade(live_spec_id) as (
+            select live_spec_id
```
As noted in a prior comment -- this is gonna do a full table scan every time.

I'm coming to believe that `controller_version` isn't buying us much and should be removed.
* We can always have migrations tweak controller_next_run if we want to do a mass-run of controllers because of a software release.
* The remaining use case for `controller_version` is understanding how to interpret a status (is it the old or new version), but we have lots of experience doing incremental updates of JSON specifications and we've never before needed a version to do so.

It seems like `controller_version`'s utility can be gotten at through other mechanisms and so it should be dropped. If that's wrong and we do find we need it, it's also easy to add it as a future migration with a default.

psFried replied:
I think `controller_version` is still needed in order to allow for agent rollouts. When we deploy a new agent version (that has an incremented `controller_version`), there will still potentially be running agent processes having an older `controller_version`. Part of the purpose of `controller_version` is to allow agents to filter controller jobs so that they don't run for rows that have already been upgraded.

TBH I was thinking that a full table scan might not be a problem in the short term. In practice, most full table scans on `live_specs` seemed to take somewhere in the ~45ms range, and I'm assuming that `controller_jobs` will be roughly similar or maybe slightly faster (fewer columns). But I haven't tested the `deque` performance under realistic conditions.

If we want to avoid the full table scans, I can think of a few approaches:
- Add an index on `controller_version`
- execute a single query when the agent starts up, which sets `controller_next_run = now()` for all rows having `controller_version < $agent_controller_version`

From my perspective, I'd _really_ like to avoid needing to do a database migration when updating the controller version. And I don't think it'd really work, given the possibility for multiple agent versions to be running simultaneously during rollouts.

psFried replied:
Following up after a VC, we decided to keep `controller_version`, but just not use it to dequeue jobs needing "upgraded". This actually allows the initial rollout of controllers for existing `live_specs` to be done incrementally. The `dequeue` function will only pick up rows that have _both_ a `controller_next_run` that's due and a corresponding `controller_jobs` row. So we can remove the initial backfill of the `controller_jobs` table from the controllers database migration, and move it into a separate migration, which is run _after_ the rollout of new agents completes. This backfilling of `controller_jobs` rows can then be done incrementally, which limits the blast radius if anything goes wrong.

psFried commented on crates/models/src/lib.rs:
```diff
@@ -42,18 +44,245 @@ pub use references::{
+        }
+    }
+
+    pub fn as_capture(&self) -> Option<&CaptureDef> {
```
The problem is that there's just _a lot_ of this type of "trivial" conversion that needs to be done, and it adds up to a ton of "noise" if we do all of them at each call site. I'm open to changing or removing these, but would ask that we bikeshed for now and take it up in a subsequent PR.

psFried commented on crates/models/src/lib.rs:
```diff
@@ -42,18 +44,245 @@ pub use references::{
     fn targets(&self) -> impl Iterator<Item = &Collection>;
+
+    /// Returns the `sources` of this spec as an owned set.
+    fn reads_from(&self) -> BTreeSet<Collection> {
```
I got rid of the separate impls, which were a holdover from before consolidating `SpecExt`. Basically, `SpecExt` and `ModelDef` were developed concurrently, and one used owned `Collection`s while the other used borrowed `&Source`s. Ideally, we'd consolidate into a single function that works for everyone, but I don't yet have a clear picture of what that should look like. See [this comment](https://github.com/estuary/flow/pull/1489#discussion_r1643403689) for more deets.

psFried commented on crates/agent-sql/src/id.rs:
```diff
@@ -5,6 +5,17 @@ use sqlx::{postgres, Decode, Encode, Type, TypeInfo};
 #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct Id([u8; 8]);
```
As it is now, we still need separate copies of `Id` and `CatalogType` because `models` doesn't depend on `sqlx` and thus can't `impl sqlx::Type`. I'm half convinced that we should add an optional `sqlx` dependency (and feature) to `models` in order to avoid this duplication and having conversions sprinkled everywhere, but figured we could address it in a subsequent PR.

psFried commented on supabase/migrations/50_controllers.sql:
```diff
@@ -0,0 +1,127 @@
+alter column flow_type set data type flow_type
+using flow_type::text::flow_type;
+
+-- We now allow live specs to be deleted even though they're still referenced by other specs.
```
Originally this was here because controllers would hard delete `live_specs` and then call `notify_dependents`, so the `live_spec_flows` were still needed in order to identify the dependent specs to notify. But now, the deletion of live specs happens _after_ the call to `notify_dependents`, and so I would think that we can keep these constraints and add `on delete cascade`. I'll give it a go

psFried replied:
Following up: the constraints have been re-added along with `on delete cascade`

psFried committed:
agent: cleanups from code review

- rename `publication_status` => `status`
- rename `publication_id` => `pub_id`
- remove unused `DraftCatalog::to_models_catalog`
- cleanup a few remaining references to stripping colons from pub ids

psFried commented on crates/agent/src/publications.rs:
```diff
@@ -1,623 +1,365 @@
+    pub built: tables::Validations,
+    /// The final status of the publication. Note that this is not neccessarily `Success`,
+    /// even if there are no `errors`.
+    pub publication_status: JobStatus,
```
done

psFried commented on crates/agent-sql/src/publications.rs:
```diff
@@ -1,10 +1,165 @@
+    pub last_pub_id: Id,
+}
+
+/// Locks the given live specs rows and returns their current revisions. This is used for verifying the revisions
```
done

jgraettinger commented on crates/agent/src/draft.rs:
```diff
@@ -1,16 +1,49 @@
+impl From<LockFailure> for Error {
+    fn from(err: LockFailure) -> Self {
+        let detail = format!(
+            "the expectVersionId of spec {:?} {:?} did not match that of the live spec {:?}",
```
nit: expectPubId?

psFried replied:
done

jgraettinger commented on crates/agent/src/handlers.rs:
```diff
@@ -113,16 +116,39 @@ enum Status {
+const PUBLICATIONS_TABLE: &str = "publications";
+const CONTROLLER_JOBS_TABLE: &str = "controller_jobs";
+
+/// The purpose of this is to ensure that the controllers handler gets polled
```
I understand our intention is to rip all of this out when we tackle the updated lease-based job queue model.

We'll still want a mechanism to immediately wake a job dequeuer from a polling backoff sleep though :thinking:

psFried replied:
Yeah, one thing I realized is that our existing notifications mechanism doesn't really work for `controller_jobs` because they don't have a `{"type":"queued"}` status, so they rely on the periodic triggers from within the agent, as well as this hack. But I think that a periodic trigger seems totally fine, as long as we can also trigger controllers promptly in response to a publication. The main improvements I'd like to see here are to make the periodic trigger more explicit and targeted to controllers, and to make this publications hack somewhat less "hacky" (perhaps :wave: by passing a `ControllersWaker` into the `Publisher`).

psFried committed:
agent: remove unused fields from publication status

The publications `JobStatus` no longer needs to include
`linked_materializations`, and the `ReCreateReason` no longer needs the
`PrevDeletedSpec` variant because we now permit re-creating things with
the same name as a previously deleted entity.

psFried committed:
agent: re-add live_spec_flows foreign keys

Turns out we don't actually need to get rid of the foreign key
constraints on `live_spec_flows`, so this adds them back in with the
addition of `on delete cascade` so that they get cleaned up when
`live_specs` rows are deleted.

jgraettinger reviewed (commented):
Made it through "agent: cleanups from code review"

A lot of my comments are unedited and stream of conscious and calling it as I see it. Perhaps we sync up to review them and discuss next steps.

This looks awesome though -- clearly a lot of work and I'm excited about what this factoring / architecture buys us.

psFried commented on crates/agent/src/controllers/collection.rs:
```diff
@@ -0,0 +1,255 @@
+        let after_minutes = now
+            .signed_duration_since(start_time)
+            .num_minutes()
+            .max(min_backoff_minutes)
```
I think this is actually right?
Starting with the duration `d` since the previous update of the inferred schema, if `d < min_backoff_minutes` then we want to use `min_backoff_minutes`. But if `d > 120 minutes`, then we take the minimum value of `120`.

psFried commented on crates/agent/src/controllers/publication_status.rs:
```diff
@@ -0,0 +1,317 @@
+                0..=1 => 60,
+                2 => 180,
+                3 => 300,
+                _ => 600,
```
No, if it has `abort` then the controller returns an error and it uses the logic [defined here](https://github.com/estuary/flow/pull/1489/files#diff-1ae46c5f49923434028707845bdf48fca06dc2f5b877dac192641a7f4893422bR122), which uses a _much_ steeper backoff. I'll admit that this separation does feel a bit half-baked, and it seems like we should be able to consolidate these cases. I'll look into it and report back.

psFried committed:
agent: update controller_version handling

Updates the controllers `dequeue` function to no longer identify rows
needing "upgraded" from an older `controller_version`, and fixes a bug
where it wasn't properly filtering out rows having a
`controller_version` that's greater than the `CONTROLLER_VERSION` of the
agent.

psFried committed:
agent: introduce controllers backfill migration

Introduces a separate database migration for incrementally backfilling
`controller_jobs` rows. The aim is to allow the initial rollout of
controllers for existing specs to be slow and incremental. This works by
inserting only a chunk (by default 1000) of rows at a time. The agent
tolerates `controller_jobs` rows that are missing, and simply skips over
any live specs that don't have a corresponding `controller_jobs` row.
So, we can add `controller_jobs` rows incrementally in order to enable
controllers for existing specs.

Note that any specs that are published by the new agent version will
automatically have controllers enabled for them. This is compatible
with the behavior of the migration, and provides a separate mechanism
for incrementally enabling controllers.

psFried committed:
agent: fix publications skipping new tests

`Publisher::build` was incorrectly checking for the presence of tests in
the _live_ catalog when determining whether tests should be run. This
meant that it would skip running tests when the tests were initially
published. This is resolved by checking for _built_ tests instead.

psFried committed:
agent: rename expectVersionId => expectPubId in error message

jgraettinger committed:
validation: omit last_version when zero instead of sending "00000000"

psFried committed:
agent: cleanup controller error handling and backoffs

Revises the approach to error handling that's used for controllers.

The `error` column of `controller_jobs` is the most obvious place to put
errors that should be shown to users, and so we should use that instead
of relying on other fields internal to the status. So controllers should
continue to just return an `anyhow::Error` and have that populate the
`error` column in most common scenarios. If there's some sort of problem
communicating with the database, the `ControlPlane` functions will
return errors, and controllers now generally respond by panicking so
that the errors are obvious to us. The exception to that rule is data
plane activations. If a call to activate something fails, we should not
crash the agent, and should instead set the error column and schedule a
retry.

The `PublicationStatus::next_run` function was removed. If a publication
fails, then the controllers return `Err`s, which get rendered into the
error column. The `RetryableError` type was added, along with a helper
trait so that controllers can use `some_result.with_retry(next_run)?` to
indicate that an error is retryable. By default, all errors are terminal
and don't result in a `controller_next_run` being set. But controllers
can explicitly opt in to retry errors when appropriate. The end result
is that the controllers are the one place where we decide how errors
should be handled, which is hopefully much more clear to readers.

psFried commented on crates/agent/src/controllers/publication_status.rs:
```diff
@@ -0,0 +1,317 @@
+        self.pending.is_some()
+    }
+
+    pub async fn start_spec_update<'a>(
```
We spoke on VC and decided to kick the can on actually doing this, but agreed that it would be a good refactor.

jgraettinger reviewed (approved):
LGTM :tada:
